src/collection.rs

```diff
@@ -7,86 +7,60 @@
 
 use error::SsError;
 use item::Item;
+use proxy::collection::CollectionProxy;
+use proxy::service::ServiceProxy;
 use session::Session;
 use ss::{
     SS_DBUS_NAME,
-    SS_INTERFACE_COLLECTION,
-    SS_INTERFACE_SERVICE,
     SS_ITEM_LABEL,
     SS_ITEM_ATTRIBUTES,
-    SS_PATH,
 };
 use util::{
     exec_prompt,
     format_secret,
-    Interface,
+    lock_or_unlock,
+    LockAction,
 };
 
-use dbus::{
-    BusName,
-    Connection,
-    MessageItem,
-    Path,
-};
-use dbus::Interface as InterfaceName;
-use dbus::MessageItem::{
-    Array,
-    Bool,
-    DictEntry,
-    ObjectPath,
-    Str,
-    Variant,
-};
-use std::rc::Rc;
-
-// Helper enum for
-// locking and unlocking
-enum LockAction {
-    Lock,
-    Unlock,
-}
+use std::collections::HashMap;
+use std::convert::TryInto;
+use zvariant::{Dict, ObjectPath, OwnedObjectPath, Value};
 
 // Collection struct.
 // Should always be created from the SecretService entry point,
 // whether through a new collection or a collection search
-#[derive(Debug, Clone)]
 pub struct Collection<'a> {
-    // TODO: Implement method for path?
-    bus: Rc<Connection>,
+    conn: zbus::Connection,
     session: &'a Session,
-    pub collection_path: Path,
-    collection_interface: Interface,
-    service_interface: Interface,
+    pub collection_path: OwnedObjectPath,
+    collection_interface: CollectionProxy<'a>,
+    service_interface: &'a ServiceProxy<'a>,
 }
 
 impl<'a> Collection<'a> {
-    pub fn new(bus: Rc<Connection>, session: &'a Session, collection_path: Path) -> Self {
-        let collection_interface = Interface::new(
-            bus.clone(),
-            BusName::new(SS_DBUS_NAME).unwrap(),
-            collection_path.clone(),
-            InterfaceName::new(SS_INTERFACE_COLLECTION).unwrap()
-        );
-        let service_interface = Interface::new(
-            bus.clone(),
-            BusName::new(SS_DBUS_NAME).unwrap(),
-            Path::new(SS_PATH).unwrap(),
-            InterfaceName::new(SS_INTERFACE_SERVICE).unwrap()
-        );
-        Collection {
-            bus,
+    pub fn new(
+        conn: zbus::Connection,
+        session: &'a Session,
+        service_interface: &'a ServiceProxy,
+        collection_path: OwnedObjectPath,
+        ) -> ::Result<Self>
+    {
+        let collection_interface = CollectionProxy::new_for_owned(
+            conn.clone(),
+            SS_DBUS_NAME.to_owned(),
+            collection_path.to_string(),
+            )?;
+        Ok(Collection {
+            conn: conn.clone(),
```

Comment:
This doesn't need cloned since `conn` is already owned and can be moved into the struct.